service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryExtractor.java

```diff
@@ -0,0 +1,232 @@
+        Predicate<Resource> containsPractitioner = RESOURCE_HAS_PRACTITIONER.getOrDefault(resource.getResourceType(), (resource1) -> false);
+        return containsPractitioner.test(resource);
+    }
+
+    private static IIdType extractIIdTypes(Resource resource) {
+        Function<Resource, IIdType> extractor = RESOURCE_EXTRACT_IIDTYPE.get(resource.getResourceType());
+        return extractor.apply(resource);
+    }
+
+    private static boolean performerContainersOrgAndPractitioner(Observation observation) {
+        var performers = observation.getPerformer();
+        return listHasPractitioner(performers) && listHasOrganization(performers);
+    }
+
+    private static boolean listHasPractitioner(List<Reference> references) {
+        return listHasResourceType(references, ResourceType.Practitioner);
+    }
+
+    private static boolean listHasOrganization(List<Reference> references) {
+        return listHasResourceType(references, ResourceType.Organization);
+    }
+
+    private static boolean listHasResourceType(List<Reference> references, ResourceType resourceType) {
+        return references.stream()
+            .map(Reference::getReferenceElement)
+            .map(IIdType::getResourceType)
+            .anyMatch(resourceType.name()::equals);
+    }
+
+    private static boolean requesterContainsAgentAndOrganization(ReferralRequest referralRequest) {
+        var requester = referralRequest.getRequester();
+        return requesterAgentContainsPractitioner(requester) && requester.hasOnBehalfOf();
+    }
+
+    private static boolean requesterAgentContainsPractitioner(ReferralRequest.ReferralRequestRequesterComponent requester) {
+        return requester.getAgent().getReferenceElement().getResourceType().equals(ResourceType.Practitioner.name());
+    }
+
+    private static boolean isPatientResource(Resource resource) {
+        return resource.getResourceType().equals(ResourceType.Patient);
+    }
+
+    private static boolean isNhsNumberMatching(Patient patient, String nhsNumber) {
+        return patient.getIdentifier()
+            .stream()
+            .filter(identifier -> identifier.getSystem().equals(NHS_NUMBER_SYSTEM))
+            .anyMatch(identifier -> identifier.getValue().equals(nhsNumber));
+    }
+
+    private static IIdType extractIIdTypeFromAllergyIntolerance(Resource resource) {
+        Optional<IIdType> practitionerReference = extractAsserterReference(resource);
+        return practitionerReference.orElseGet(() -> ((AllergyIntolerance) resource).getRecorder().getReferenceElement());
+    }
+
+    private static Optional<IIdType> extractAsserterReference(Resource resource) {
+        var practitionerReference = ((AllergyIntolerance) resource).getAsserter().getReferenceElement();
+        if (practitionerReference.getIdPart() != null) {
+            return Optional.of(practitionerReference);
+        }
+        return Optional.empty();
+    }
```

Comment:
Since this is expecting an `AllergyIntolerance`, the parameter could reasonably be that specific. The calling method can cast. Also, can use `filter` instead of the conditional branch if you like

```suggestion
    private static Optional<IIdType> extractAsserterReference(AllergyIntolerance allergyIntolerance) {
        var practitionerReference = allergyIntolerance.getAsserter().getReferenceElement();
        return Optional.of(practitionerReference)
            .filter(ref -> ref.getIdPart() != null);
    }
```